ddtrace/contrib/dogpile_cache/patch.py

```diff
@@ -0,0 +1,38 @@
+import dogpile
+
+from ddtrace.pin import Pin, _DD_PIN_NAME, _DD_PIN_PROXY_NAME
+from ddtrace.vendor.wrapt import wrap_function_wrapper as _w
+
+from .lock import _wrap_lock_ctor
+from .region import _wrap_get_create, _wrap_get_create_multi
+
+_get_or_create = dogpile.cache.region.CacheRegion.get_or_create
+_get_or_create_multi = dogpile.cache.region.CacheRegion.get_or_create_multi
+_lock_ctor = dogpile.lock.Lock.__init__
+
+
+def patch():
+    if getattr(dogpile.cache, '_datadog_patch', False):
+        return
+    setattr(dogpile.cache, '_datadog_patch', True)
+
+    _w('dogpile.cache.region', 'CacheRegion.get_or_create', _wrap_get_create)
+    _w('dogpile.cache.region', 'CacheRegion.get_or_create_multi', _wrap_get_create_multi)
+    # Used to track lock acquisitions and releases.
+    _w('dogpile.lock', 'Lock.__init__', _wrap_lock_ctor)
+
+    Pin(app='dogpile.cache', service='dogpile.cache', app_type='CACHE').onto(dogpile.cache)
+
+
+def unpatch():
+    if not getattr(dogpile.cache, '_datadog_patch', False):
+        return
+    setattr(dogpile.cache, '_datadog_patch', False)
+    # This looks silly but the unwrap util doesn't support class instance methods, even
+    # though wrapt does. This was causing the patches to stack on top of each other
+    # during testing.
+    dogpile.cache.region.CacheRegion.get_or_create = _get_or_create
+    dogpile.cache.region.CacheRegion.get_or_create_multi = _get_or_create_multi
+    dogpile.lock.Lock.__init__ = _lock_ctor
```

Comment:
This is super-nitpicky but there's a possibility of:
1. This module gets loaded
2. Some other patching occurs
3. DD patching occurs
4. DD unpatching
and the last step would wipe out the patching in step 2 because the original methods are saved at import time. Not sure if that level of rigor is necessary since unpatch() seems most useful for tests but just FYI

Reply:
Yeah the alt is to fix their unwrap method which I don't want to do right now. And yeah it seems it's only used for testing. We'll see if they push back.